stdlib/src/collections/dict.mojo

```diff
@@ -939,17 +943,22 @@ struct Dict[K: KeyElement, V: CollectionElement](
             if self._over_compact_factor():
                 self._compact()
             return
-        var _reserved = self._reserved() * 2
-        self.size = 0
-        self._n_entries = 0
-        var old_entries = self._entries^
-        self._entries = self._new_entries(_reserved)
-        self._index = _DictIndex(self._reserved())
+        var _reserved = self._reserved()
 
-        for i in range(len(old_entries)):
-            var entry = old_entries.__get_ref(i)
-            if entry[]:
-                self._insert(entry[].unsafe_take())
+        for i in range(_reserved):
+            self._entries.append(None)
```

Comment:
This feels off for a couple reasons.
- Why are we doubling the size when we're over the compaction factor, but not over the load factor? That's the wrong behavior.
- As far as I can tell, in the resize case, this will optimistically be the same performance, assuming that `List` always doubles in size which is exactly what we want to do here; however if that weren't the case this would actually do extra copies of `None` values. Is there any benefit I'm missing? If not this is just harder to understand to me vs. manually creating and copying the new list.

Reply:
Hello,

The first check for compact factor is done on line 1 of _maybe_resize;
it is the same as before (only done when were not overloaded, and returns).

If it does not return, we double the size,
the second check for compact factor, after the resize, can be removed.
it should never get done, and is there just in case.
(could be turned into a debug_assert)

Even a full list of size self._reserved(), doubled,
would not be at compact factor of self._reserved()*2 :

size 8, _n_entries 8, reserved 8 (impossible to reach, would been doubled !)

doubled reserved 16:
(8 * 4) > (16 * 3) 
(32 > 48) == False

This is why when we are overloaded, we are technically never overcompact,
because we have to double the size anyway.
So it does not compact on overload, taking into consideration the new reserved.

(Part of the performance come from not compacting on overload)

&nbsp;

For the list replacing, 
`_new_entries` first create a list full of none and return it,
then every existing entries are moved in to replace some.
(Self.reserved * 2 moves of none, +  self.size move of non-none)

here, `List._realloc` move half the new size with previous values, and we add none.
(Self.reserved * moves of none and non-none, +  Self.reserved move of none)

(without replacing values)

So both implementation archeive 2*Self.reserved, but PR don't replace values.
(`List.append` initialize self._reserved() items with none, without replacing)

Not sure if it is better or not, but not `List.__del__` too !